frontend/app/src/next/components/waterfall/waterfall.tsx

```diff
@@ -0,0 +1,467 @@
+          isExpanded: expandedTasks.has(task.metadata.id),
+          // Chart data
+          offset: (queuedAt - globalMinQueuedAt) / 1000, // in seconds
+          queuedDuration: (startedAt - queuedAt) / 1000, // in seconds
+          ranDuration: (finishedAt - startedAt) / 1000, // in seconds
+          status: task.status,
+          taskId: task.taskId, // Add taskId for tie-breaking in sorting
+        };
+      })
+      .filter((task): task is NonNullable<typeof task> => task !== null)
+      // Sort by task path for consistent ordering, break ties with taskId
+      .sort((a, b) => {
+        const pathA = taskPathMap.get(a.id) || [];
+        const pathB = taskPathMap.get(b.id) || [];
+
+        // Compare each path segment
+        for (let i = 0; i < Math.min(pathA.length, pathB.length); i++) {
+          if (pathA[i] !== pathB[i]) {
+            const taskA = taskMap.get(pathA[i]);
+            const taskB = taskMap.get(pathB[i]);
+
+            if (taskA && taskB) {
+              return taskA.taskId - taskB.taskId;
+            }
+            return 0;
+          }
+        }
+
+        // If one path is a prefix of the other, shorter comes first
+        if (pathA.length !== pathB.length) {
+          return pathA.length - pathB.length;
+        }
+
+        // If paths are identical, break ties with taskId
+        return a.taskId - b.taskId;
+      });
+
+    return { data, taskPathMap };
+  }, [taskData, expandedTasks, depth, autoExpandedInitially]);
+
+  // Custom tick renderer with expand/collapse buttons
+  const renderTick = (props: {
+    x: number;
+    y: number;
+    payload: { value: string };
+  }) => {
+    const { x, y, payload } = props;
+    const task = processedData.data.find(
+      (t) => t.taskDisplayName === payload.value,
+    );
+    if (!task) {
+      // Return empty element instead of null
+      return <g transform={`translate(${x},${y})`}></g>;
+    }
+
+    const label = payload.value;
+    const truncatedLabel =
+      label.length > 16 ? label.slice(0, 16) + '...' : label;
+    const indentation = task.depth * 12; // 12px indentation per level
+
+    return (
+      <g transform={`translate(${x},${y})`}>
+        <foreignObject
+          x={-160} // Start position (right aligned)
+          y={-10} // Vertically center
+          width={160}
+          height={20}
+        >
+          <div
+            style={{
+              display: 'flex',
+              alignItems: 'center',
+              paddingLeft: `${indentation}px`,
+              height: '100%',
+            }}
+          >
+            {/* Expand/collapse button */}
+            <div
+              style={{
+                cursor: task.hasChildren ? 'pointer' : 'default',
+                display: 'flex',
+                alignItems: 'center',
+                justifyContent: 'center',
+                width: '20px',
+                height: '20px',
+                marginRight: '4px',
```

Comment:
tailwind here?

Reply:
For some reason, some of these components didn't seem to be applying tailwind styles, perhaps because it's all rendered in an SVG canvas. I can try it again though.

Reply:
ooo yeah that's definitely possible! not a huge deal